src/lib/uri/uri.c

```diff
@@ -456,3 +459,32 @@ uri_unescape(const char *src, size_t src_size, char *dst, bool decode_plus)
 	}
 	return (size_t)(dst_buf - dst);
 }
+
+/* Compare two strings which might be NULL. */
+static bool
+fields_are_equal(const char *a, const char *b)
+{
+	if ((a == NULL) != (b == NULL))
+		return false;
+	if (a == NULL)
+		return true;
+	return strcmp(a, b) == 0;
+}
+
+bool
+uri_equals(const struct uri *a, const struct uri *b)
+{
+	/*
+	 * Either service or path will be NULL depending on whether this is a
+	 * unix socket or not.
+	 */
+	return fields_are_equal(a->host, b->host) &&
+	       fields_are_equal(a->path, b->path) &&
+	       fields_are_equal(a->service, b->service);
+}
+
+bool
+uri_is_nil(const struct uri *uri)
+{
+	return memcmp(uri, &uri_nil, sizeof(*uri)) == 0;
```

Comment:
This won't work for a uri which has all members 0, but non-zero padding between the members. Maybe better check just `host` member? Without it the uri has no purpose anyway.

Reply:
> Maybe better check just host member? Without it the uri has no purpose anyway.

Not really, for example:
```
tarantool> require('uri').parse(3301)
---
- service: '3301'
...

```

But checking for `host`, `service` and `path` should work just fine. Done:
```diff
diff --git a/src/lib/uri/uri.c b/src/lib/uri/uri.c
index 7a162c266..a67c2ff64 100644
--- a/src/lib/uri/uri.c
+++ b/src/lib/uri/uri.c
@@ -486,5 +486,5 @@ uri_equals(const struct uri *a, const struct uri *b)
 bool
 uri_is_nil(const struct uri *uri)
 {
-       return memcmp(uri, &uri_nil, sizeof(*uri)) == 0;
+       return uri->host == NULL && uri->path == NULL && uri->service == NULL;
 }
```

Reply:
After giving it some thought, I'd like to return to checking for a fully empty uri.
The reasons are: the name uri_is_nil() implies that the uri is really unset / empty. But if we check for host / path / service, some strange uris (like "#fragment", "?q1=v1") and such will still be considered nil. Which's not fine IMO.

Besides, 'uri_create(NULL)` does a memset to 0, so such a uri won't have a non-zero padding.